core_impl/src/main/java/com/google/instrumentation/trace/SpanFactoryImpl.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.trace;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.instrumentation.common.Clock;
+import com.google.instrumentation.trace.Link.Type;
+import com.google.instrumentation.trace.Span.Options;
+import com.google.instrumentation.trace.TraceConfig.TraceParams;
+import java.util.EnumSet;
+import java.util.List;
+import java.util.Random;
+import javax.annotation.Nullable;
+
+/** Implementation of the {@link SpanFactory}. */
+final class SpanFactoryImpl extends SpanFactory {
+  private static final EnumSet<Span.Options> noRecordEventOptions =
+      EnumSet.noneOf(Span.Options.class);
+  private static final EnumSet<Span.Options> recordEventOptions = EnumSet.of(Options.RECORD_EVENTS);
+
+  private final RandomHandler randomHandler;
+  private final SpanImpl.StartEndHandler startEndHandler;
+  private final Clock clock;
+  private final TraceConfig traceConfig;
+
+  @Override
+  protected Span startSpan(@Nullable Span parent, String name, StartSpanOptions options) {
+    SpanContext parentContext = null;
+    TimestampConverter timestampConverter = null;
+    if (parent != null) {
+      parentContext = parent.getContext();
+      // Pass the timestamp converter from the parent to ensure that the recorded events are in
+      // the right order. Implementation uses System.nanoTime() which is monotonically increasing.
+      if (parent instanceof SpanImpl) {
+        timestampConverter = ((SpanImpl) parent).getTimestampConverter();
+      }
+    }
+    return startSpanInternal(parentContext, false, name, options, timestampConverter);
+  }
+
+  @Override
+  protected Span startSpanWithRemoteParent(
+      @Nullable SpanContext remoteParent, String name, StartSpanOptions options) {
+    return startSpanInternal(remoteParent, true, name, options, null);
+  }
+
+  Span startSpanInternal(
+      @Nullable SpanContext parent,
+      boolean hasRemoteParent,
+      String name,
+      StartSpanOptions startSpanOptions,
+      TimestampConverter timestampConverter) {
+    TraceParams activeTraceParams = traceConfig.getActiveTraceParams();
+    Random random = randomHandler.current();
+    TraceId traceId;
+    SpanId spanId = SpanId.generateRandomId(random);
+    SpanId parentSpanId = null;
+    TraceOptions.Builder traceOptionsBuilder;
+    if (parent == null || !parent.isValid()) {
```

Comment:
So currently if the user explicitly set the option for this to become a root we pass null, but if no current Span in the context and the user does not specify the parent (or force to be a root) the context is invalid and we will pass the invalid context, or the last option is if the user cannot parse a receiving context from the wire they may pass invalid context instead of null. 

I know it is horrible and maybe I should change the API to always pass null.